### PR TITLE
fix: infer from usage duplicate

### LINKED ISSUE #44059: Refactoring `Infer type from usage` duplicate function type
TS stable & canary
[Playground Link](https://www.typescriptlang.org/play?noImplicitAny=false#code/DYUwLgBAzgrgRgbgLAChWzgfgHQAoCUyKGEAvBAWQHwDeAvkA)
```ts
let sub;

sub?.();
sub = () => {}
```
chrome_2021-05-12_12-12-07

Output :
```ts
let sub: { (): void; (): void; };
```

Expected:
```ts
let sub: () => void;
```

### PR DESCRIPTION
**Pull request by ZYSzys**
Fixes #44059

### CODE REVIEW

File: src/services/codefixes/inferFromUsage.ts [hunk 1]
```diff
@@ -930,7 +930,19 @@ namespace ts.codefix {
                     low: t => !!(getObjectFlags(t) & ObjectFlags.Anonymous)
                 }];
             let good = removeLowPriorityInferences(inferences, priorities);
-            const anons = good.filter(i => getObjectFlags(i) & ObjectFlags.Anonymous) as AnonymousType[];
+
+            const anons = [];
+            const displayParts: string[] = [];
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts:935]: Using an array + indexOf is not a great way to dedupe. Map/Set would be better

File: src/services/codefixes/inferFromUsage.ts [hunk 2]
```diff
@@ -930,7 +930,19 @@ namespace ts.codefix {
...
                 }];
             let good = removeLowPriorityInferences(inferences, priorities);
-            const anons = good.filter(i => getObjectFlags(i) & ObjectFlags.Anonymous) as AnonymousType[];
+
+            const anons = [];
+            const displayParts: string[] = [];
+            // filter same displayPart type
+            for (const item of good) {
+                const isAnonymousType = getObjectFlags(item) & ObjectFlags.Anonymous;
+                const displayPart = item.checker.typeToString(item);
```
Comment (sandersn) [inline src/services/codefixes/inferFromUsage.ts:939]: I don't like using strings as the key for de-duping. It happens pretty late and means that we're not addressing the cause of the dupes. Why are the dupes happening? Maybe it's possible to construct the types in a way that eliminates them without a de-duping pass.

**Review by sandersn — CHANGES_REQUESTED** [review]

**Comment by sandersn** [comment]
@ZYSzys Do you have time to find the root cause of this bug? If not I'd like to close the PR to keep the number of open PRs manageable.

**Comment by ZYSzys** [comment]
> @ZYSzys Do you have time to find the root cause of this bug? If not I'd like to close the PR to keep the number of open PRs manageable.

@sandersn Maybe I don't have too much time to investigate it deeply, it's ok to close the PR.

**Pull request closed without merging**